### DISCUSSION TITLE
Multi-file project support

### DISCUSSION THREAD

**Discussion by evilguest** [volarjs/volar.js#187] [Q&A] [answered]
Hello everyone.
I am trying to apply VolarJS for the DSL designed by the company I work for. The language itself is pretty elaborate - types, functions, async messaging, etc - and features the usual concept of modules.
So far I've got the VS Code client and the langage server, stemming from the volarjs/starter project. The language parsing is built via the [Ohm](https://ohmjs.org/) project and works pretty well. 
Everything works reasonably fine within a single language file - I am able to provide the diagnostics, definitions, references, and hover content. 

However, in order for the extension to really work, I need a way to "peek" into the other files of the project. 
And that's where I've got stuck.
Two major issues:
1. Within my ServicePluginInstance, in most request I do just have the TextDocument instance, and can get to the VirtualCode from it. How do I get to the VirtualCode instances for the neighbour files there? 
2. When responding to provideDiagnostics, I might stumble on an error in the referenced file. However, the Diagnostic object contains just the range. How do I report the problems in the files beyond the current one being served? 

Would appreciate any help here, since all the samples I've found so far are just doing some magic for Typescript; and the language I'm designing this for is not Typescript.

**Comment by johnsoncodehk** [comment]
1. You can use the context param from `LanguageServicePlugin#create` function.

```ts
const myLSPlugin: LanguageServicePlugin = {
	create(context) {
		return {
			provideXXX(document) {
				const dir = Utils.dirname(URI.parse(document.uri));
				const files = await context.env.fs?.readDirectory(dir) ?? [];
				if (files.length) {
					const otherScriptUri = Utils.joinPath(dir, files[0][0]);
					const otherVirtualCode = context.language.scripts.get(otherScriptUri)?.generated?.root;
					// ...
				}
			},
		};
	},
};
```

2. I don't fully understand this use case, but it looks a bit like what relatedInformation does.

```ts
const myLSPlugin: LanguageServicePlugin = {
	create(context) {
		return {
			provideDiagnostics(document) {
				return [{
					relatedInformation: [{
						// ...
					}],
				}];
			},
		};
	},
};
```

**Reply by evilguest** [reply]
Thanks a lot for the response.
1. Looks like the context.language.scripts is a recent addition in 2.2.x. Added it, struggled a bit with decoding the new schema of the document.uri, but now the overall thing seems to be working.
2. I've applied an approach you've suggested. The way I've managed it is to report an error for the "import xxx" statement in the "main" document, referring to all the diagnostics found in the referenced document. However, this is a bit different than what I've wanted to achieve. 
Idea is to be able to show "project-wide" errors separately, even if the problematic file is not open in any of the editor windows. 
In an extreme case the user just does have the _folder_ with the project open; is there a way for me to proactively scan the folder and push the found issues to the IDE _before_ the user ever opens any of the files? 
3. Extra obstacle within the current solution: if I have both the "main" and "referenced" documents open in the IDE, and there is an issue in the "referenced" file, there is a diagnostic generated in the "main" file (and, maybe, some other secondary problems that are caused by the "referenced" code - i.e. some of the functions does have a bug, therefore the "main" file fails to find its definition).
Now I fix the "referenced" code, so it is no longer broken. The document being edited gets automatically re-validated. However, the dependent document does not get the re-validation, and still shows the outdated diagnostics.
The question is: how do I force VSCode to re-validate all the dependent files?

**Reply by johnsoncodehk** [reply, marked as answer]
2. You need to use WorkspaceDiagnostics, which has been added via #199.
3. You need to replace `provideDiagnostics` with `provideSemanticDiagnostics`. `provideSemanticDiagnostics` will fire when any other file changes.

**Comment by evilguest** [comment]
Thanks a lot, #3 works like charm. Will review the #2 promptly.

Anton Zlygostev
________________________________
От: Johnson Chu ***@***.***>
Отправлено: Sunday, June 9, 2024 5:51:50 AM
Кому: volarjs/volar.js ***@***.***>
Копия: Anton Zlygostev ***@***.***>; Author ***@***.***>
Тема: Re: [volarjs/volar.js] Multi-file project support (Discussion #187)


  1.  You need to use WorkspaceDiagnostics, which has been added via #199<https://github.com/volarjs/volar.js/pull/199>.
  2.  You need to replace provideDiagnostics with provideSemanticDiagnostics. provideSemanticDiagnostics will fire when any other file changes.
